### ISSUE TITLE
Reboot "Software engineering principles" course

### ISSUE THREAD

**Issue by astrojuanlu** [kedro-org/kedro-devrel#110]
After delivering several editions of the original "Software engineering principles" course #13 we got generally good feedback 🎉  and an ample majority of beginners felt that the course was worth their time (~90 %). However, we also observed

- A large drop-off ratio, and
- Consistent feedback that the course was too basic

Before resuming these trainings, I'd like to iterate on the syllabus and possibly cover fewer topics in the same amount of time (because the course is already quite long, 6 hours in total). The good thing is that we have a recorded version internally that we can always refer to.

Some places we can draw inspiration from:

- "Software Engineering for Data Scientists" by Catherine Nelson, to be released next year https://www.oreilly.com/library/view/software-engineering-for/9781098136192/
- Ian Ozsvald "Software Engineering for Data Scientists" trainings https://ianozsvald.com/training/software-engineering-for-data-scientists/
- Ploomber course on building ETL pipelines https://ploomber-sql.readthedocs.io/en/latest/packaging-your-sql-project/intro-to-python-scripting-and-pipelines.html

Some ideas:

- More gradual editor journey, starting on Jupyter, then transitioning to VSCode notebooks, then using the full IDE capabilities?
- Less focus on GitHub as a platform (better introduced elsewhere) and more on real world git workflows?
  - Or, alternatively, remove the git CLI completely and use Jupyter and VSCode graphical interfaces instead? (It can be argued that a full 6 hour course could be devoted just to git, and in just a few minutes of play we are not setting participants for success)
- A single coherent story rather than disjoint exercises for the different sections?
- Less focus on basic aspects of Python functions and more real-world refactoring exercises?
- Less focus on YAML the language and YAML vs JSON?

**Comment by stichbury** [comment]
* I'm a reviewer on Cat's book (as is Antony) so we have visibility on that.
* When you say "fewer topics in the same time" doesn't that mean you are spending longer on them, or doing more depth? Because the former means the course gets easier, while the latter means it gets more focussed but still kinda basic.

**Comment by astrojuanlu** [comment]
Good catch, I was thinking of more depth, rather than just streching them
